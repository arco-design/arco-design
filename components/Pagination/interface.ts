import { CSSProperties, ReactNode } from 'react';
import { SelectProps } from '../Select/interface';

/**
 * @title Pagination
 */
export interface PaginationProps {
  style?: CSSProperties;
  className?: string | string[];
  /**
   * @zh 分页按钮样式
   * @en Pagination button style
   */
  pageItemStyle?: CSSProperties;
  /**
   * @zh 被选中的分页按钮样式
   * @en The style of the selected page button
   */
  activePageItemStyle?: CSSProperties;
  /**
   * @zh 当前页
   * @en Current page
   */
  current?: number;
  /**
   * @zh 每页数据条数
   * @en Number of data items per page
   */
  pageSize?: number;
  /**
   * @zh 数据总数
   * @en Total number of data
   */
  total?: number;
  /**
   * @zh 当前页默认值
   * @en To set default current page
   */
  defaultCurrent?: number;
  /**
   * @zh 默认每页数据条数
   * @en To set default number of data per page
   */
  defaultPageSize?: number;
  /**
   * @zh 是否禁用
   * @en Whether to disable
   */
  disabled?: boolean;
  /**
   * @zh 是否在只有一页的情况下隐藏
   * @en Whether to hide when there is only one page
   * @version 2.6.0
   */
  hideOnSinglePage?: boolean;
  /**
   * @zh 定制分页按钮的结构
   * @en Customized pagination button structure
   */
  itemRender?: (
    page: number,
    type: 'page' | 'more' | 'prev' | 'next',
    originElement: ReactNode
  ) => ReactNode;
  /**
   * @zh 分页器尺寸
   * @en pager size
   */
  size?: 'mini' | 'small' | 'default' | 'large';
  /**
   * @zh 是否显示数据总数
   * @en Whether to display the total number of data
   */
  showTotal?: boolean | ((total: number, range: number[]) => ReactNode);
  /**
   * @zh 是否可以改变每页条数
   * @en Is it possible to change page size
   * @defaultValue true
   */
  sizeCanChange?: boolean;
  /**
   * @zh 每页可以显示数据条数
   * @en The number of data items that can be displayed per page
   */
  sizeOptions?: number[];
  /**
   * @zh `current` 页与 `...` 之间的页码个数
   * @en the number of pages between the `current` page and `...`
   * @defaultValue 2
   * @version 2.32.0
   */
  bufferSize?: number;
  /**
   * @zh 变化时的回调
   * @en Callback when page changes
   */
  onChange?: (pageNumber: number, pageSize: number) => void;
  /**
   * @zh pageSize 变化时的回调
   * @en Callback when pageSize changes
   */
  onPageSizeChange?: (size: number, current: number) => void;
  /**
   * @zh `pageSize` 改变的时候重置当前页码为 `1`
   * @en When pageSize changes, resets the current page number to `1`
   * @defaultValue true
   */
  pageSizeChangeResetCurrent?: boolean;
  /**
   * @zh 是否应用精简分页模式
   * @en Whether to use simplified pagination mode
   */
  simple?: boolean;
  /**
   * @zh 是否显示快速跳转到某页，在 `simple` 模式下默认为 true
   * @en Whether to display quick jump. Defaults to true in `simple` mode
   */
  showJumper?: boolean;
  /**
   * @zh 是否显示更多页码提示（当尚无法计算数据总数时可以使用）
   * @en Whether to show more page number tips (can be used when the total number of data cannot be calculated yet)
   */
  showMore?: boolean;
  /**
   * @zh 用于配置弹出框的属性
   * @en Props of the `Select`
   */
  selectProps?: Partial<SelectProps>;
  /**
   * @zh 设置分页器的图标
   * @en Set icon of the pager
   */
  icons?: {
    prev?: ReactNode;
    next?: ReactNode;
    more?: ReactNode;
  };
  mini?: boolean; // 1.0
}
