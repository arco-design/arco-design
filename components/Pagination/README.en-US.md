`````
Component / Navigation

# Pagination

Use pagination to control the amount of information in a single page.
`````

%%Content%%

## API

### Pagination

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|style|Additional style|`CSSProperties`|`-`|-|
|className|Additional css class|`string \| string[]`|`-`|-|
|pageItemStyle|Pagination button style|`CSSProperties`|`-`|-|
|activePageItemStyle|The style of the selected page button|`CSSProperties`|`-`|-|
|current|Current page|`number`|`-`|-|
|pageSize|Number of data items per page|`number`|`-`|-|
|total|Total number of data|`number`|`-`|-|
|defaultCurrent|To set default current page|`number`|`-`|-|
|defaultPageSize|To set default number of data per page|`number`|`-`|-|
|disabled|Whether to disable|`boolean`|`-`|-|
|hideOnSinglePage|Whether to hide when there is only one page|`boolean`|`-`|2.6.0|
|itemRender|Customized pagination button structure|`(page: number,type: 'page' \| 'more' \| 'prev' \| 'next',originElement: ReactNode) => ReactNode`|`-`|-|
|size|pager size|`'mini' \| 'small' \| 'default' \| 'large'`|`-`|-|
|showTotal|Whether to display the total number of data|`boolean \| ((total: number, range: number[]) => ReactNode)`|`-`|-|
|sizeCanChange|Is it possible to change page size|`boolean`|`true`|-|
|sizeOptions|The number of data items that can be displayed per page|`number[]`|`-`|-|
|onChange|Callback when page changes|`(pageNumber: number, pageSize: number) => void`|`-`|-|
|onPageSizeChange|Callback when pageSize changes|`(size: number, current: number) => void`|`-`|-|
|pageSizeChangeResetCurrent|When pageSize changes, resets the current page number to `1`|`boolean`|`true`|-|
|simple|Whether to use simplified pagination mode|`boolean`|`-`|-|
|showJumper|Whether to display quick jump|`boolean`|`-`|-|
|showMore|Whether to show more page number tips (can be used when the total number of data cannot be calculated yet)|`boolean`|`-`|-|
|selectProps|Props of the `Select`|`Partial<SelectProps>`|`-`|-|
|icons|Set icon of the pager|`{prev?: ReactNode;next?: ReactNode;more?: ReactNode;}`|`-`|-|
