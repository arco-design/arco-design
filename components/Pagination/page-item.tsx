import React, { useContext, ReactNode, CSSProperties } from 'react';
import cs from '../_util/classNames';
import IconLeft from '../../icon/react-icon/IconLeft';
import IconRight from '../../icon/react-icon/IconRight';
import IconMore from '../../icon/react-icon/IconMore';
import { ConfigContext } from '../ConfigProvider';
import useKeyboardEvent from '../_util/hooks/useKeyboardEvent';

type itemRenderType = (
  page: number,
  type: 'page' | 'more' | 'prev' | 'next',
  originElement: ReactNode
) => ReactNode;

export interface PagerProps {
  pageItemStyle?: CSSProperties;
  activePageItemStyle?: CSSProperties;
  disabled?: boolean;
  rootPrefixCls: string;
  pageNum: number;
  current: number;
  onClick: (value: number) => void;
  itemRender?: itemRenderType;
}

export interface JumpPagerProps {
  pageItemStyle?: CSSProperties;
  disabled?: boolean;
  rootPrefixCls: string;
  type: StepType;
  current: number;
  allPages: number;
  jumpPage: number;
  onClick: (value: number) => void;
  icons?: {
    prev?: ReactNode;
    next?: ReactNode;
    more?: ReactNode;
  };
  itemRender?: itemRenderType;
}

export enum StepType {
  previous,
  next,
}
export interface StepPagerProps {
  pageItemStyle?: CSSProperties;
  disabled?: boolean;
  rootPrefixCls: string;
  type: StepType;
  current: number;
  allPages: number;
  onClick: (value: number) => void;
  icons?: {
    prev?: ReactNode;
    next?: ReactNode;
    more?: ReactNode;
  };
  itemRender?: itemRenderType;
}

/**
 * default pager item
 */

function Pager(props: PagerProps) {
  const { locale } = useContext(ConfigContext);
  const getKeyboardEvents = useKeyboardEvent();

  const { pageNum, current, rootPrefixCls, pageItemStyle, activePageItemStyle, itemRender } = props;

  const prefixCls = `${rootPrefixCls}-item`;
  const isActive = current === pageNum;
  const classnames = cs(prefixCls, isActive ? `${prefixCls}-active` : '');

  let style = pageItemStyle;
  if (isActive) {
    style = { ...style, ...activePageItemStyle };
  }

  const ariaCurrentProps = isActive ? { 'aria-current': true } : {};

  const onClick = (e) => {
    const { pageNum, onClick, disabled } = props;
    if (e.currentTarget.dataset.active === 'true') {
      return;
    }

    e.stopPropagation();

    if (!disabled) {
      onClick && onClick(pageNum);
    }
  };

  return (
    <li
      style={style}
      className={classnames}
      onClick={onClick}
      tabIndex={props.disabled ? -1 : 0}
      aria-label={locale.Pagination.currentPage?.replace('{0}', pageNum)}
      {...getKeyboardEvents({ onPressEnter: onClick })}
      {...ariaCurrentProps}
    >
      {itemRender ? itemRender(pageNum, 'page', pageNum) : pageNum}
    </li>
  );
}

function getIcon(name: string, icons) {
  switch (name) {
    case 'prev':
      return icons && icons.prev ? icons.prev : <IconLeft />;
    case 'next':
      return icons && icons.next ? icons.next : <IconRight />;
    case 'more':
      return icons && icons.more ? icons.more : <IconMore />;
    default:
      return null;
  }
}

/**
 * preJump or nextJump button
 * @param props
 */
export const JumpPager = (props: JumpPagerProps) => {
  const { locale } = useContext(ConfigContext);

  const { rootPrefixCls, current, allPages, jumpPage, icons, disabled, pageItemStyle, itemRender } =
    props;

  const minCurrent = allPages > 0 ? 1 : 0;
  const nextPage = Math.min(allPages, Math.max(minCurrent, current + jumpPage));
  const prefix = `${rootPrefixCls}-item ${rootPrefixCls}-item-jumper`;

  const cls = cs(prefix);
  const onClick = () => {
    !disabled && props.onClick && props.onClick(nextPage);
  };

  const originElement = getIcon('more', icons);

  const ariaLabel =
    jumpPage > 0
      ? locale.Pagination.nextSomePages?.replace('{0}', jumpPage)
      : locale.Pagination.prevSomePages?.replace('{0}', -jumpPage);

  return (
    <li style={pageItemStyle} className={cls} onClick={onClick} aria-label={ariaLabel}>
      {itemRender ? itemRender(undefined, 'more', originElement) : originElement}
    </li>
  );
};

/**
 * previous or next button
 * @param props
 */
export const StepPager = (props: StepPagerProps) => {
  const { locale, rtl } = useContext(ConfigContext);

  const getKeyboardEvents = useKeyboardEvent();
  const { rootPrefixCls, current, allPages, type, icons, disabled, pageItemStyle, itemRender } =
    props;
  const prefixCls = `${rootPrefixCls}-item`;
  const [prev, next] = rtl ? ['next', 'prev'] : ['prev', 'next'];
  const StepIcon = type === StepType.previous ? getIcon(prev, icons) : getIcon(next, icons);
  let _disabled = false;
  if (allPages === 0) {
    // total为0
    _disabled = true;
  } else if (type === StepType.previous) {
    // 向前翻页
    _disabled = current <= 1; // current ===0 || current===1
  } else {
    // 向后翻页
    _disabled = current === allPages;
  }
  const innerDisabled = disabled || _disabled;

  let nextPage = current + (type === StepType.previous ? -1 : 1);
  nextPage = Math.max(0, Math.min(allPages, nextPage));

  const pageType = StepType.previous === type ? 'prev' : 'next';

  const cls = cs(prefixCls, `${prefixCls}-${pageType}`, {
    [`${prefixCls}-disabled`]: innerDisabled,
  });
  const onClick = () => {
    if (innerDisabled) {
      return;
    }
    props.onClick && props.onClick(nextPage);
  };

  return (
    <li
      style={pageItemStyle}
      className={cls}
      onClick={onClick}
      tabIndex={innerDisabled ? -1 : 0}
      aria-label={locale.Pagination[pageType]}
      {...getKeyboardEvents({
        onPressEnter: onClick,
      })}
    >
      {itemRender ? itemRender(undefined, pageType, StepIcon) : StepIcon}
    </li>
  );
};

export default Pager;
