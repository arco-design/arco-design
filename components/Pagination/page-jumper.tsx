import React, { useRef, useState, useContext, useEffect } from 'react';
import Input from '../Input';
import { isFunction, isUndefined, isObject } from '../_util/is';
import { ConfigContext } from '../ConfigProvider';
import { RefInputType } from '../Input/interface';

export interface PageJumperProps {
  disabled?: boolean;
  rootPrefixCls?: string;
  totalPages: number;
  current: number;
  simple?: boolean | { showJumper?: boolean };
  onPageChange?: (value) => void;
  size?: 'mini' | 'small' | 'default' | 'large';
}

function PageJumper(props: PageJumperProps) {
  const defaultInputText = props.simple ? props.current : undefined;
  const { locale } = useContext(ConfigContext);
  const [inputText, setInputText] = useState<number>(defaultInputText);
  const inputRef = useRef<RefInputType>();

  useEffect(() => {
    if (props.simple) {
      setInputText(props.current);
    }
  }, [props.simple, props.current]);

  const handleChange = (val) => {
    const value = parseInt(val, 10);
    setInputText(isNaN(value) ? undefined : value);
  };

  const handleJump = () => {
    const { onPageChange, totalPages, current, simple } = props;

    if (isUndefined(inputText)) {
      return;
    }

    if (inputText === current && !simple) {
      setInputText(undefined);
      return;
    }

    let page = isNaN(Number(inputText)) ? current : Number(inputText);
    if (page < 1) {
      page = 1;
    } else if (page > totalPages) {
      page = totalPages;
    }

    setInputText(simple ? page : undefined);
    isFunction(onPageChange) && (onPageChange as Function)(page);
  };

  const onFocus = () => {
    const input = inputRef.current.dom;
    if (String(inputText) && input) {
      input.setSelectionRange(0, String(inputText).length);
    }
  };

  const { rootPrefixCls, totalPages, simple, size, disabled } = props;
  const prefixCls = `${rootPrefixCls}-jumper`;
  const inputConfig = { showJumper: true, ...(isObject(simple) ? simple : {}) };

  return (
    <div className={`${prefixCls}`}>
      {!simple && <span className={`${prefixCls}-text-goto`}>{locale.Pagination.goto}</span>}
      {inputConfig.showJumper ? (
        <Input
          _ignorePropsFromGlobal
          ref={(ref) => (inputRef.current = ref)}
          className={`${prefixCls}-input`}
          value={!isUndefined(inputText) ? inputText.toString() : undefined}
          size={size}
          disabled={disabled || !totalPages}
          onChange={handleChange}
          onPressEnter={handleJump}
          onFocus={onFocus}
          onBlur={handleJump}
        />
      ) : (
        <span>{inputText}</span>
      )}
      {!simple && <span className={`${prefixCls}-text-goto-suffix`}>{locale.Pagination.page}</span>}
      {simple && (
        <>
          <span className={`${prefixCls}-separator`}>/</span>
          <span>{totalPages}</span>
        </>
      )}
    </div>
  );
}

export default PageJumper;
