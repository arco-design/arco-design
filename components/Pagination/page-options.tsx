import React, { useRef, useContext } from 'react';
import Select from '../Select';
import { ConfigContext } from '../ConfigProvider';
import { SelectProps } from '../Select/interface';

const noop = () => {};
const Option = Select.Option;
const _SizeOptions = [10, 20, 30, 40, 50];

export interface PageOptionProps {
  disabled?: boolean;
  rootPrefixCls?: string;
  sizeCanChange?: boolean;
  sizeOptions?: number[];
  onPageSizeChange?: (value) => void;
  pageSize?: number;
  size?: 'mini' | 'small' | 'default' | 'large';
  selectProps?: Partial<SelectProps>;
}

function PageOption(props: PageOptionProps) {
  const selectRef = useRef();
  const { locale } = useContext(ConfigContext);
  const {
    sizeCanChange = false,
    onPageSizeChange = noop,
    rootPrefixCls,
    sizeOptions = _SizeOptions,
    pageSize = 10,
    size,
    selectProps,
    disabled,
  } = props;

  return (
    sizeCanChange && (
      <div
        ref={selectRef}
        className={`${rootPrefixCls}-option`}
        aria-label={locale.Pagination.pageSize}
      >
        <Select
          value={sizeOptions.indexOf(pageSize) !== -1 ? pageSize : sizeOptions[0]}
          onChange={(value) => {
            onPageSizeChange(value);
          }}
          size={size}
          getPopupContainer={() => selectRef.current}
          disabled={disabled}
          {...selectProps}
        >
          {sizeOptions.map((num) => {
            return (
              <Option key={num} value={num}>
                {`${num} ${locale.Pagination.countPerPage}`}
              </Option>
            );
          })}
        </Select>
      </div>
    )
  );
}

export default PageOption;
