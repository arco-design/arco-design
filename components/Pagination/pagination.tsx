import React, { ReactNode, ReactElement, useEffect, useContext, forwardRef } from 'react';
import PageItem, { StepType, JumpPager, StepPager } from './page-item';
import useMergeValue from '../_util/hooks/useMergeValue';
import PageOption from './page-options';
import PageJumper from './page-jumper';
import cs from '../_util/classNames';
import { ConfigContext } from '../ConfigProvider';
import { PaginationProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';
import { pickDataAttributes } from '../_util/pick';

export interface PaginationState {
  current: number;
  pageSize: number;
  total?: number;
  showMore?: boolean;
}

const _defaultCurrent = 1;
const _defaultPageSize = 10;

function getAllPages(pageSize = _defaultPageSize, total) {
  return Math.ceil(total / pageSize);
}

function getBufferSize(bufferSize, allPages) {
  const min = 0;
  const max = Math.floor(allPages / 2) - 1;
  const newBufferSize = Math.max(bufferSize, min);
  return Math.min(newBufferSize, max);
}

function getAdjustPageSize(sizeOptions?: number[]) {
  if (sizeOptions && sizeOptions.length) {
    return sizeOptions[0];
  }
  return _defaultPageSize;
}

const defaultProps: PaginationProps = {
  total: 0,
  pageSizeChangeResetCurrent: true,
  bufferSize: 2,
};

function Pagination(baseProps: PaginationProps, ref) {
  const { getPrefixCls, size: ctxSize, locale, componentConfig, rtl } = useContext(ConfigContext);
  const props = useMergeProps<PaginationProps>(
    baseProps,
    defaultProps,
    componentConfig?.Pagination
  );
  const {
    total: propTotal,
    pageSize: propPageSize,
    current: propCurrent,
    showMore: propShowMore,
    sizeOptions: propSizeOptions,
    pageSizeChangeResetCurrent,
    defaultCurrent,
    defaultPageSize,
  } = props;

  const [current, setCurrent] = useMergeValue(_defaultCurrent, {
    defaultValue: defaultCurrent,
    value: propCurrent,
  });

  const [pageSize, setPageSize] = useMergeValue(getAdjustPageSize(propSizeOptions), {
    defaultValue: defaultPageSize,
    value: propPageSize,
  });

  const total = propTotal;
  const showMore = !!propShowMore;

  if (propCurrent && !props.onChange) {
    console.warn(
      'Warning: you have provide current prop for pagination but without onChange handler ,' +
        ' this will cause no-change when you change page. '
    );
  }

  function getAdjustedCurrent(newPageSize, newTotal) {
    const newAllPages = getAllPages(newPageSize, newTotal);
    const newCurrent = current > newAllPages ? newAllPages : current;
    return newCurrent;
  }

  useEffect(() => {
    // adjust pageSize after sizeOption changes
    const needAdjust = propSizeOptions && !propSizeOptions.includes(pageSize);
    // trigged when currentPageSize not in the options;
    if (needAdjust) {
      const adjustPageSize = getAdjustPageSize(propSizeOptions);
      if (!('pageSize' in props)) {
        setPageSize(adjustPageSize);
      }
    }
  }, [propSizeOptions]);

  useEffect(() => {
    // adjust currentPage after total and pageSize changes
    const newCurrent = getAdjustedCurrent(pageSize, total);
    if (newCurrent !== current && !('current' in props)) {
      setCurrent(newCurrent);
    }
  }, [total, current, pageSize]);

  const onChange = (pageNumber = current, size = pageSize) => {
    const { onChange } = props;
    onChange && onChange(pageNumber, size);
  };

  const onPageSizeChange = (pageSize) => {
    const { onPageSizeChange } = props;
    const allPages = getAllPages(pageSize, total);
    const newState = {
      pageSize,
    } as PaginationState;
    if (pageSizeChangeResetCurrent) {
      newState.current = 1;
    } else {
      newState.current = current > allPages ? allPages : current;
    }
    if (!('pageSize' in props)) {
      setPageSize(newState.pageSize);
    }

    if (!('current' in props) && current !== newState.current) {
      setCurrent(newState.current);
    }

    onPageSizeChange && onPageSizeChange(pageSize, newState.current);
    onChange(pageSizeChangeResetCurrent ? 1 : newState.current, pageSize);
  };

  const onPageNumberChange = (pageNumber) => {
    if (!('current' in props)) {
      setCurrent(pageNumber);
    }
    onChange(pageNumber);
  };

  const {
    className,
    style,
    pageItemStyle,
    activePageItemStyle,
    showTotal,
    sizeCanChange,
    sizeOptions,
    simple,
    mini,
    showJumper,
    selectProps,
    icons,
    disabled,
    itemRender,
    hideOnSinglePage,
  } = props;

  const size = props.size || ctxSize;
  const prefixCls = getPrefixCls('pagination');
  // compatible 1.0
  const innerSize = mini ? 'mini' : size;
  const classNames = cs(
    prefixCls,
    `${prefixCls}-size-${innerSize}`,
    {
      [`${prefixCls}-simple`]: simple,
      [`${prefixCls}-disabled`]: disabled,
      [`${prefixCls}-rtl`]: rtl,
    },
    className
  );

  let renderPager: ReactElement;
  const pageList: ReactElement[] = [];
  const allPages = getAllPages(pageSize, total);

  const bufferSize = getBufferSize(props.bufferSize, allPages);

  if (hideOnSinglePage && allPages <= 1) {
    return null;
  }

  const pagerProps = {
    onClick: onPageNumberChange,
    rootPrefixCls: prefixCls,
    simple,
    current,
    allPages,
    icons,
    disabled,
    pageItemStyle,
    activePageItemStyle,
    itemRender,
  };

  // simple mode, no pager list
  if (simple) {
    const prefix = `${prefixCls}-item-simple`;
    renderPager = (
      <ul className={`${prefixCls}-list`}>
        <StepPager key="previous" {...pagerProps} type={StepType.previous} />
        <li className={`${prefix}-pager`}>
          <PageJumper
            disabled={disabled}
            rootPrefixCls={prefixCls}
            totalPages={allPages}
            current={current}
            onPageChange={onPageNumberChange}
            simple={{ showJumper: typeof showJumper === 'boolean' ? showJumper : true }}
            size={innerSize}
          />
        </li>
        <StepPager key="next" {...pagerProps} type={StepType.next} />
      </ul>
    );
  } else {
    // fold = ... >= 2pages;
    const beginFoldPage = 1 + 2 + bufferSize;
    const endFoldPage = allPages - 2 - bufferSize;
    if (
      // beginPage(1 page) + bufferSize * 2 + endPage(1 page) + ...(2 pages)
      allPages <= 4 + bufferSize * 2 ||
      (current === beginFoldPage && current === endFoldPage)
    ) {
      for (let i = 1; i <= allPages; i++) {
        pageList.push(<PageItem {...pagerProps} key={i} pageNum={i} />);
      }
    } else {
      let left = 1;
      let right = allPages;
      let hasJumpPre = true;
      let hasJumpNext = true;

      // fold front and back
      if (current > beginFoldPage && current < endFoldPage) {
        right = current + bufferSize;
        left = current - bufferSize;
        // fold back
      } else if (current <= beginFoldPage) {
        hasJumpPre = false;
        left = 1;
        right = Math.max(beginFoldPage, bufferSize + current);
        // fold begin
      } else if (current >= endFoldPage) {
        hasJumpNext = false;
        right = allPages;
        left = Math.min(endFoldPage, current - bufferSize);
      }

      for (let i = left; i <= right; i++) {
        pageList.push(<PageItem key={i} pageNum={i} {...pagerProps} />);
      }
      const JumpPre = (
        <JumpPager
          {...pagerProps}
          key={left - 1}
          type={StepType.previous}
          jumpPage={-(bufferSize * 2 + 1)}
        />
      );
      const JumpNext = (
        <JumpPager
          {...pagerProps}
          key={right + 1}
          type={StepType.next}
          jumpPage={bufferSize * 2 + 1}
        />
      );
      const FirstPager = <PageItem key={1} pageNum={1} {...pagerProps} />;
      const LastPager = <PageItem {...pagerProps} key={allPages} pageNum={allPages} />;
      if (hasJumpPre) {
        pageList[0] = React.cloneElement(pageList[0], {
          className: `${prefixCls}-item-after-pre`,
        });
        // TODO:
        pageList.unshift(JumpPre);
        pageList.unshift(FirstPager);
      }

      if (hasJumpNext) {
        pageList[pageList.length - 1] = React.cloneElement(pageList[pageList.length - 1], {
          className: `${prefixCls}-item-before-next`,
        });
        pageList.push(JumpNext);
        pageList.push(LastPager);
      }
    }

    renderPager = (
      <ul className={`${prefixCls}-list`}>
        <StepPager {...pagerProps} key="previous" type={StepType.previous} />
        {pageList}
        {showMore && (
          <JumpPager
            {...pagerProps}
            key={allPages + 1}
            type={StepType.next}
            jumpPage={bufferSize * 2 + 1}
          />
        )}
        <StepPager key="next" {...pagerProps} type={StepType.next} />
      </ul>
    );
  }

  let totalElement: ReactNode = null;
  if (typeof showTotal === 'boolean' && showTotal) {
    totalElement = (
      <div className={`${prefixCls}-total-text`}>
        {locale.Pagination.total?.replace('{0}', total)}
      </div>
    );
  }
  if (typeof showTotal === 'function') {
    totalElement = (
      <div className={`${prefixCls}-total-text`}>
        {showTotal(total, [(current - 1) * pageSize + 1, current * pageSize])}
      </div>
    );
  }

  return (
    <div {...pickDataAttributes(props)} className={classNames} style={style} ref={ref}>
      {totalElement}
      {renderPager}
      <PageOption
        disabled={disabled}
        rootPrefixCls={prefixCls}
        sizeCanChange={sizeCanChange}
        sizeOptions={sizeOptions}
        onPageSizeChange={onPageSizeChange}
        pageSize={pageSize}
        size={innerSize}
        selectProps={selectProps}
      />
      {!simple && showJumper && (
        <PageJumper
          disabled={disabled}
          rootPrefixCls={prefixCls}
          totalPages={allPages}
          current={current}
          onPageChange={onPageNumberChange}
          size={innerSize}
        />
      )}
    </div>
  );
}

const PaginationComponent = forwardRef<unknown, PaginationProps>(Pagination);

PaginationComponent.displayName = 'Pagination';

export default PaginationComponent;

export { PaginationProps };
