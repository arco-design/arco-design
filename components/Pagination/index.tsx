import Pagination from './pagination';

export default Pagination;

export type { PaginationProps } from './interface';
