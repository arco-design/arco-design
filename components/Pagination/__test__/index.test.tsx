import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-test-renderer';
import { Enter } from '../../_util/keycode';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import Pagination from '..';
import { PaginationProps, PaginationState } from '../pagination';

mountTest(Pagination);
componentConfigTest(Pagination, 'Pagination');

function mountPagination(component: React.ReactElement) {
  return mount<React.Component, React.PropsWithChildren<PaginationProps>, PaginationState>(
    component
  );
}

describe('Pagination', () => {
  it('pagination should be update correctly', () => {
    const wrapper = mountPagination(<Pagination />);
    act(() => {
      wrapper.setProps({
        total: 100,
      });
      wrapper.update();
    });
    expect(wrapper.find('.arco-pagination-list').childAt(0).prop('allPages')).toBe(10);
  });

  it('pagination should be unmount correctly', () => {
    const wrapper = mountPagination(<Pagination />);
    expect(() => {
      wrapper.unmount();
    }).not.toThrow();
  });

  it('pagination should jump correctly pages', () => {
    const wrapper = mountPagination(<Pagination total={200} showJumper />);
    expect(wrapper.find('.arco-pagination-item-active').text()).toBe('1');
    wrapper
      .find('PageJumper')
      .find('Input')
      .simulate('change', { target: { value: 30 } });
    wrapper.find('PageJumper').find('Input').simulate('keyDown', { keyCode: Enter.code });
    expect(wrapper.find('.arco-pagination-item-active').text()).toBe('20');
  });

  it('should change pageSize correctly', () => {
    const wrapper = mountPagination(<Pagination total={200} showTotal sizeCanChange />);
    wrapper.setProps({ pageSize: 30 });
    expect(wrapper.find('.arco-pagination-item')).toHaveLength(Math.ceil(200 / 30) + 2);
    wrapper.setProps({
      total: 0,
      pageSize: 10,
    });
    wrapper.setProps({ pageSize: 30 });
    expect(wrapper.find('.arco-pagination-item')).toHaveLength(2);
    expect(wrapper.find('.arco-pagination-item-disabled')).toHaveLength(2);
  });

  it('trigger onPageSizeChange correctly', () => {
    const mockPageSizeChange = jest.fn();
    const mockChange = jest.fn();
    const wrapper = mountPagination(
      <Pagination
        total={200}
        onChange={mockChange}
        onPageSizeChange={mockPageSizeChange}
        sizeCanChange
        sizeOptions={[10, 20, 50]}
        pageSizeChangeResetCurrent={false}
      />
    );
    wrapper.find('Pager').last().simulate('click');
    expect(mockChange.mock.calls[0]).toEqual([200 / 10, 10]);

    wrapper.find('PageOption').find('Select').simulate('click');
    wrapper.find('PageOption').find('.arco-select-option').last().simulate('click');
    expect(mockChange.mock.calls[1]).toEqual([200 / 50, 50]);
    expect(mockPageSizeChange.mock.calls[0]).toEqual([50, 200 / 50]);

    wrapper.setProps({
      pageSizeChangeResetCurrent: true,
    });

    wrapper.find('PageOption').find('Select').simulate('click');
    wrapper.find('PageOption').find('.arco-select-option').first().simulate('click');
    expect(mockChange.mock.calls[2]).toEqual([1, 10]);
    expect(mockPageSizeChange.mock.calls[1]).toEqual([10, 1]);
  });
});
