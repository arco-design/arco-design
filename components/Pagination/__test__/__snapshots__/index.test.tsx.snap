// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Pagination default 1`] = `
<div
  class="arco-pagination arco-pagination-size-default"
>
  <ul
    class="arco-pagination-list"
  >
    <li
      aria-label="上一页"
      class="arco-pagination-item arco-pagination-item-prev arco-pagination-item-disabled"
    >
      <svg
        aria-hidden="true"
        class="arco-icon arco-icon-left"
        fill="none"
        focusable="false"
        stroke="currentColor"
        stroke-width="4"
        viewBox="0 0 48 48"
      >
        <path
          d="M32 8.4 16.444 23.956 32 39.513"
        />
      </svg>
    </li>
    <li
      aria-label="下一页"
      class="arco-pagination-item arco-pagination-item-next arco-pagination-item-disabled"
    >
      <svg
        aria-hidden="true"
        class="arco-icon arco-icon-right"
        fill="none"
        focusable="false"
        stroke="currentColor"
        stroke-width="4"
        viewBox="0 0 48 48"
      >
        <path
          d="m16 39.513 15.556-15.557L16 8.4"
        />
      </svg>
    </li>
  </ul>
</div>
`;

exports[`ConfigProvider componentConfig.Pagination set className globally 1`] = `
<div
  class="arco-pagination arco-pagination-size-default global-default-class"
>
  <ul
    class="arco-pagination-list"
  >
    <li
      aria-label="上一页"
      class="arco-pagination-item arco-pagination-item-prev arco-pagination-item-disabled"
    >
      <svg
        aria-hidden="true"
        class="arco-icon arco-icon-left"
        fill="none"
        focusable="false"
        stroke="currentColor"
        stroke-width="4"
        viewBox="0 0 48 48"
      >
        <path
          d="M32 8.4 16.444 23.956 32 39.513"
        />
      </svg>
    </li>
    <li
      aria-label="下一页"
      class="arco-pagination-item arco-pagination-item-next arco-pagination-item-disabled"
    >
      <svg
        aria-hidden="true"
        class="arco-icon arco-icon-right"
        fill="none"
        focusable="false"
        stroke="currentColor"
        stroke-width="4"
        viewBox="0 0 48 48"
      >
        <path
          d="m16 39.513 15.556-15.557L16 8.4"
        />
      </svg>
    </li>
  </ul>
</div>
`;

exports[`ConfigProvider componentConfig.Pagination set className globally and component set className itself. 1`] = `
<div
  class="arco-pagination arco-pagination-size-default component-class"
>
  <ul
    class="arco-pagination-list"
  >
    <li
      aria-label="上一页"
      class="arco-pagination-item arco-pagination-item-prev arco-pagination-item-disabled"
    >
      <svg
        aria-hidden="true"
        class="arco-icon arco-icon-left"
        fill="none"
        focusable="false"
        stroke="currentColor"
        stroke-width="4"
        viewBox="0 0 48 48"
      >
        <path
          d="M32 8.4 16.444 23.956 32 39.513"
        />
      </svg>
    </li>
    <li
      aria-label="下一页"
      class="arco-pagination-item arco-pagination-item-next arco-pagination-item-disabled"
    >
      <svg
        aria-hidden="true"
        class="arco-icon arco-icon-right"
        fill="none"
        focusable="false"
        stroke="currentColor"
        stroke-width="4"
        viewBox="0 0 48 48"
      >
        <path
          d="m16 39.513 15.556-15.557L16 8.4"
        />
      </svg>
    </li>
  </ul>
</div>
`;
